tensorflow_quantum/python/optimizers/rotosolve_minimizer.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2020 The TensorFlow Quantum Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""The rotosolve minimization algorithm
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import math
+import collections
+import tensorflow as tf
+
+
+def prefer_static_shape(x):
+    """Return static shape of tensor `x` if available,
+     else `tf.shape(x)`.
+    Args:
+      x: `Tensor` (already converted).
+    Returns:
+      Numpy array (if static shape is obtainable), else `Tensor`.
+    """
+    return prefer_static_value(tf.shape(x))
+
+
+def prefer_static_value(x):
+    """Return static value of tensor `x` if available, else `x`.
+    Args:
+      x: `Tensor` (already converted).
+    Returns:
+      Numpy array (if static value is obtainable), else `Tensor`.
+    """
+    static_x = tf.get_static_value(x)
+    if static_x is not None:
+        return static_x
+    return x
+
+
+RotosolveOptimizerResults = collections.namedtuple(
+    'RotosolveOptimizerResults',
+    [
+        'converged',  # Scalar boolean tensor indicating whether the minimum
+        # was found within tolerance.
+        'num_iterations',  # The number of iterations of the BFGS update.
+        # The total number of objective evaluations performed.
+        'num_objective_evaluations',
+        'position',  # A tensor containing the last argument value found
+        # during the search. If the search converged, then
+        # this value is the argmin of the objective function.
+        # A tensor containing the value of the objective from previous iteration
+        'last_objective_value',
+        'objective_value',  # A tensor containing the value of the objective
+        # function at the `position`. If the search
+        # converged, then this is the (local) minimum of
+        # the objective function.
+        'tolerance',  # Define the stop criteria. Iteration will stop when the
+        # objective value difference between two iterations is smaller than
+        # tolerance
+        'solve_param_i',  # The parameter index where rotosolve is currently
+        # modifying. Reserved for internal use.
+    ])
+
+
+def _get_initial_state(initial_position, tolerance, expectation_value_function):
+    """Create RotosolveOptimizerResults with initial state of search ."""

```

Comment:
please remove white space between the last character and period.

Reply:
Fixed in 564256fa0316a94cc5fc82a6285ebe64285afab8